extract condition struct

jwsui opened the pull request:
Extract CRD condition struct so that all CRDs could share the same condition struct. Update CRD SecurityPolicy to use the common condition struct instead of defining by itselt.

heypnus commented on pkg/apis/v1alpha1/condition.go:
```diff
@@ -0,0 +1,22 @@
+	NotReady ConditionType = "NotReady"
+)
+
+// Condition defines condition of CRD.
```
How about `Condition defines the condition of the custom resource.`? Because I think it's a condition of CR, not CRD.

jwsui replied:
Indeed, fixed.

lxiaopei commented on pkg/apis/v1alpha1/condition.go:
```diff
@@ -0,0 +1,22 @@
+
+const (
+	Ready    ConditionType = "Ready"
+	NotReady ConditionType = "NotReady"
```
Ready status has value "True or False", do we still need "NotReady" Status?

jwsui replied:
Indeed, removed `NotReady` as we don't need it anymore, thanks.

lxiaopei commented on pkg/apis/v1alpha1/condition.go:
```diff
@@ -0,0 +1,22 @@
+	Type ConditionType `json:"type"`
+	// Status defines status of condition type, True or False.
+	Status corev1.ConditionStatus `json:"status"`
+	// Reason shows a brief reason of condition.
```
Could we also add this field?

// Last time the condition transitioned from one status to another.
	// This should be when the underlying condition changed. If that is not known, then using the time when
	// the API field changed is acceptable.
	// +required
	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`

jwsui replied:
Added, but a minor question, we have kubebuilder label `+required` which means it's a a required field while we also have `omitempty` in json tag which indicate that it's option. Do we have conflict here？

lxiaopei replied:
sorry, I think we can set // +optional for it.

lxiaopei commented on pkg/apis/v1alpha1/condition.go:
```diff
@@ -0,0 +1,22 @@
+type Condition struct {
+	// Type defines condition type.
+	Type ConditionType `json:"type"`
+	// Status defines status of condition type, True or False.
```
Status of the condition, one of True, False, Unknown.

jwsui replied:
Thanks, fixed.

lxiaopei commented on pkg/apis/v1alpha1/securitypolicy_types.go:
```diff
@@ -110,31 +110,9 @@ type SecurityPolicyPort struct {
-	// Ready indicates whether NSX Operator was able to successfully make
-	// a PATCH call to create/update Security Policy on NSX associated
-	// with the user created Security Policy CR
-	SecurityPolicyReady SecurityPolicyStatusCondition = "Ready"
```
SecurityPolicyReady  is used in securitypolicy_controller.go, if removed here, need to change that file.

jwsui replied:
My bad, didn't notice that we already have controller code for securitypolicy CRD. Updated `securitypolicy_controller.go` and `securitypolicy_controller_test.go`, thanks for reminder.

lxiaopei commented on pkg/apis/v1alpha1/condition.go:
```diff
@@ -0,0 +1,29 @@
+package v1alpha1
```
could we rename the file as condition_types.go?

jwsui replied:
sure, fixed.

lxiaopei reviewed (commented):
Did you try to generate binary using the changed files?

jwsui commented:
> Did you try to generate binary using the changed files?

Yes, generate binary successfully by `make build`.

jianjuns commented:
Please make sure the commit message contains enough information for readers to understand the change. I changed the PR title to: "Extract common CRD Status Condition struct"

jwsui commented:
> Please make sure the commit message contains enough information for readers to understand the change. I changed the PR title to: "Extract common CRD Status Condition struct"

Thanks Jianjun, updated the commit message.

lxiaopei commented:
LGTM

lxiaopei reviewed (approved):
LGTM

jwsui commented:
PR merged, thanks for help review.
